pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2024 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package infracluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	mapiv1beta1 "github.com/openshift/api/machine/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	cerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ptr "k8s.io/utils/ptr"
+	azurev1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+var (
+	errUnableToGetAzureMAPIProviderSpec = errors.New("unable to get Azure MAPI ProviderSpec")
+)
+
+const (
+	clusterSecretName               = "capz-manager-cluster-credential"    // #nosec G101
+	capzManagerBootstrapCredentials = "capz-manager-bootstrap-credentials" // #nosec G101
+)
+
+// ensureAzureCluster ensures the AzureCluster cluster object exists.
+//
+//nolint:funlen
+func (r *InfraClusterController) ensureAzureCluster(ctx context.Context, log logr.Logger) (client.Object, error) {
+	target := &azurev1.AzureCluster{ObjectMeta: metav1.ObjectMeta{
+		Name:      r.Infra.Status.InfrastructureName,
+		Namespace: defaultCAPINamespace,
+	}}
+
+	// Checking whether InfraCluster object exists. If it doesn't, create it.
+	if err := r.Get(ctx, client.ObjectKeyFromObject(target), target); err != nil && !cerrors.IsNotFound(err) {
+		return nil, fmt.Errorf("failed to get InfraCluster: %w", err)
+	} else if err == nil {
+		return target, nil
+	}
+
+	log.Info(fmt.Sprintf("AzureCluster %s/%s does not exist, creating it", target.Namespace, target.Name))
+
+	// Get the Azure Bootstrap Credentials Secret.
+	capzManagerBootstrapSecret := &corev1.Secret{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: defaultCAPINamespace, Name: capzManagerBootstrapCredentials}, capzManagerBootstrapSecret); err != nil {
+		if !cerrors.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to get Azure Boostrap Credentials Secret: %w", err)
+		}
+	}
+
+	// Get the Azure Cluster Secret.
```

Comment:
What's the purpose of this secret?

Reply:
We're attempting to get the `capzManagerBootstrapSecret` [from here](https://github.com/openshift/cluster-capi-operator/blob/47e02d0f326bbf65714f5b6f400995c332ea0e2b/manifests/0000_30_cluster-api_01_credentials-request.yaml#L78), and if it isn't found with the contents we expect, we then attempt to create this secret instead. I think this should answer your question above too if I'm not mistaken :)

Reply:
Hmm, that's not how I'm reading this.

If the CAPZ secret exists, we get to this point and create the cluster secret using data from it?

Where is this secret read? It seems to be referenced in the AzureClusterIdentity?

Are we copying this secret across because it has a different format requirement to what CCO produces?

Reply:
If I'm not mistaken, I think Joel here is referring to why we are creating a new secret to reference within the AzureIdentity rather that referencing directly the CAPZ Manager Bootstrap Secret, which gets created by the CCO.

The reason lies in the structure of the Secret's Data. CAPZ controllers expect a [specific Data structure](https://capz.sigs.k8s.io/topics/identities#service-principal) that is different to what CCO provides.

Reply:
Ahhh yes sorry Joel I misunderstood you. Thank you for the explanation Dam!

Reply:
Can we explain that in a comment with the code please, I fear that may get lost with time

Reply:
Done!